rfcs/2021-12-20-10517-llvm-backend-for-vrl.md

```diff
@@ -0,0 +1,560 @@
+  call void @vrl_expression_op_eq_impl(%"std::result::Result<vrl_compiler::Value, vrl_compiler::ExpressionError>"* %rhs, %"std::result::Result<vrl_compiler::Value, vrl_compiler::ExpressionError>"* %result)
+  call void @vrl_resolved_drop(%"std::result::Result<vrl_compiler::Value, vrl_compiler::ExpressionError>"* %rhs)
+  %vrl_resolved_boolean_is_true = call i1 @vrl_resolved_boolean_is_true(%"std::result::Result<vrl_compiler::Value, vrl_compiler::ExpressionError>"* %result)
+  br i1 %vrl_resolved_boolean_is_true, label %if_statement_if_branch, label %if_statement_else_branch
+
+if_statement_end:                                 ; preds = %if_statement_else_branch, %block_end
+  ret void
+
+if_statement_if_branch:                           ; preds = %literal_begin
+  br label %block_begin
+
+if_statement_else_branch:                         ; preds = %literal_begin
+  br label %if_statement_end
+
+block_begin:                                      ; preds = %if_statement_if_branch
+  br label %assignment_single_begin
+
+block_end:                                        ; preds = %block_next, %block_error
+  br label %if_statement_end
+
+block_error:                                      ; preds = %assignment_single_end
+  br label %block_end
+
+assignment_single_begin:                          ; preds = %block_begin
+  br label %literal_begin1
+
+assignment_single_end:                            ; preds = %literal_begin1
+  %vrl_resolved_is_err = call i1 @vrl_resolved_is_err(%"std::result::Result<vrl_compiler::Value, vrl_compiler::ExpressionError>"* %result)
+  br i1 %vrl_resolved_is_err, label %block_error, label %block_next
+
+literal_begin1:                                   ; preds = %assignment_single_begin
+  call void @vrl_expression_literal_impl(%"memmem::SearcherKind"* bitcast ([40 x i8]* @"\22bar\22" to %"memmem::SearcherKind"*), %"std::result::Result<vrl_compiler::Value, vrl_compiler::ExpressionError>"* %result)
+  call void @vrl_expression_assignment_target_insert_external_impl(%"vrl_compiler::Context"* %context, %"lookup_buf::LookupBuf"* bitcast ([32 x i8]* @foo to %"lookup_buf::LookupBuf"*), %"std::result::Result<vrl_compiler::Value, vrl_compiler::ExpressionError>"* %result)
+  br label %assignment_single_end
+
+block_next:                                       ; preds = %assignment_single_end
+  br label %block_end
+}
+```
+
+After running several LLVM optimization passes over the LLVM IR:
+
+```llvm
+; Function Attrs: nofree norecurse nosync nounwind readnone uwtable willreturn
+define void @vrl_execute(%142* noalias nocapture align 8 dereferenceable(32) %0, %752* noalias nocapture align 8 dereferenceable(88) %1) unnamed_addr #87 personality i32 (i32, i32, i64, %462*, %9*)* @rust_eh_personality {
+  %3 = alloca %529*, align 8
+  %4 = alloca %752, align 8
+  %5 = alloca [5 x i64], align 8
+  %6 = alloca %135, align 8
+  %7 = alloca %116, align 8
+  %8 = alloca %135, align 8
```

Comment:
These are stack allocations? Should we be seeing corresponding drops for these?

Reply:
Yes, they are. No, not necessarily, these allocations come from anywhere within the implementation of the inlined functions. The types that have been allocated there may have an empty `Drop` implementation that LLVM completely optimized out.

Reply:
Oh, and just to clarify. These are of course popped from the stack frame before exiting the function, LLVM doesn't need any explicit instruction to do that.

Reply:
I guess, thinking out loud: if we allocate on the stack via one of the FFI-exported "stdlib" functions, and similarly consume those stack allocated values via one of the stdlib functions, or pass it on as the return value... we'd be fine, but isn't Rust generating calls to the drop glue for values to make this possible?

In that vein, how do _we_ handle ensuring temporaries are "dropped" in a Rust sense?

Reply:
If you look at the unoptimized program above, you can see the following succession of instructions:

```llvm
...
%rhs = alloca %"std::result::Result<vrl_compiler::Value, vrl_compiler::ExpressionError>", align 8
call void @vrl_resolved_initialize(%"std::result::Result<vrl_compiler::Value, vrl_compiler::ExpressionError>"* %rhs)
...
call void @vrl_resolved_drop(%"std::result::Result<vrl_compiler::Value, vrl_compiler::ExpressionError>"* %rhs)
...
```

Not calling `vrl_resolved_initialize` and `vrl_resolved_drop` at appropriate times simply is undefined behavior.

We can add something to the module builder API that when you request to build a temporary value immediately calls `vrl_resolved_initialize` and inserts `vrl_resolved_drop` when being dropped from the builder scope.

If we use the [`llvm.lifetime.start`](https://llvm.org/docs/LangRef.html#llvm-lifetime-start-intrinsic) and [`llvm.lifetime.end`](https://llvm.org/docs/LangRef.html#llvm-lifetime-end-intrinsic) intrinsics, that should guard us against use of uninitialized values or usage after the value has been dropped.

Reply:
Added mitigation strategy in dacd0a8b455d2c72597c926a49badec552354b40.